contracts/MetaGuard.sol

```diff
@@ -0,0 +1,187 @@
+        setUp(initializeParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
+        __Ownable_init();
+        (address _owner, address _avatar, address[] memory _guards) = abi
+            .decode(initParams, (address, address, address[]));
+
+        avatar = _avatar;
+        setupGuards();
+        transferOwnership(_owner);
+
+        for (uint256 i = 0; i < _guards.length; i++) {
+            addGuard(_guards[i]);
+        }
+
+        emit MetaGuardSetup(msg.sender, _owner, _avatar, _guards);
+    }
+
+    function setupGuards() internal {
+        require(
+            guards[SENTINEL_GUARDS] == address(0),
+            "setupGuards has already been called"
+        );
+        guards[SENTINEL_GUARDS] = SENTINEL_GUARDS;
+    }
+
+    function setAvatar(address _avatar) public onlyOwner {
+        avatar = _avatar;
+        emit AvatarSet(avatar);
+    }
+
+    /// @dev Guard transactions only use the first four parameters: to, value, data, and operation.
+    /// Guard.sol hardcodes the remaining parameters as 0 since they are not used for guard transactions.
+    /// @notice This interface is used to maintain compatibilty with Gnosis Safe transaction guards.
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures,
+        address msgSender
+    ) external override {
+        address currentGuard = guards[SENTINEL_GUARDS];
+        while (currentGuard != SENTINEL_GUARDS) {
+            IGuard(currentGuard).checkTransaction(
+                to,
+                value,
+                data,
+                operation,
+                safeTxGas,
+                baseGas,
+                gasPrice,
+                gasToken,
+                refundReceiver,
+                signatures,
+                msgSender
+            );
+            currentGuard = guards[currentGuard];
+        }
+    }
+
+    function checkAfterExecution(bytes32 txHash, bool success)
+        external
+        override
+    {
+        address currentGuard = guards[SENTINEL_GUARDS];
+        while (currentGuard != SENTINEL_GUARDS) {
+            IGuard(currentGuard).checkAfterExecution(txHash, success);
+            currentGuard = guards[currentGuard];
+        }
+    }
+
+    /// @dev Removes a guard on meta guard.
+    /// @notice This can only be called by the owner.
+    /// @param prevGuard Guard that pointed to the guard to be removed in the linked list.
+    /// @param guard Guard to be removed.
+    function removeGuard(address prevGuard, address guard) public onlyOwner {
+        if (guard == address(0) || guard == SENTINEL_GUARDS)
+            revert InvalidGuard(guard);
+        if (guards[prevGuard] != guard) revert AlreadyRemovedGuard(guard);
```

Comment:
In case the guard is not present, and it never was, will this still throw `AlreadyRemovedGuard`? If this is the case then the error sounds a bit odd because the guard was never present (and it wasn't removed) and the message could be misleading

Reply:
yes, it will still throw `AlreadyRemovedGuard`, What if `UnknownGuard`? Is that better?